tool/tctl/common/bots_command.go

```diff
@@ -238,9 +283,138 @@ func (c *BotsCommand) AddBot(ctx context.Context, client auth.ClientI) error {
 	})
 }
 
+// AddBot adds a new certificate renewal bot to the cluster.
+func (c *BotsCommand) AddBot(ctx context.Context, client auth.ClientI) error {
+	// Jankily call the endpoint invalidly. This lets us version check and use
+	// the legacy version of this CLI tool if we are talking to an older
+	// server.
+	{
+		_, err := client.BotServiceClient().CreateBot(ctx, &machineidv1pb.CreateBotRequest{
+			Bot: nil,
+		})
+		if trace.IsNotImplemented(err) {
+			return trace.Wrap(c.addBotLegacy(ctx, client))
+		}
+	}
+
+	roles := splitRoles(c.botRoles)
+	if len(roles) == 0 {
+		log.Warning("No roles specified. The bot will not be able to produce outputs until a role is added to the bot.")
+	}
+	var token types.ProvisionToken
+	var err error
+	if c.tokenID == "" {
+		// If there's no token specified, generate one
+		tokenName, err := utils.CryptoRandomHex(16)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		ttl := c.tokenTTL
+		if ttl == 0 {
+			ttl = defaults.DefaultBotJoinTTL
+		}
+		tokenSpec := types.ProvisionTokenSpecV2{
+			Roles:      types.SystemRoles{types.RoleBot},
+			JoinMethod: types.JoinMethodToken,
+			BotName:    c.botName,
+		}
+		token, err = types.NewProvisionTokenFromSpec(tokenName, time.Now().Add(ttl), tokenSpec)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if err := client.UpsertToken(ctx, token); err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		// If there is, check the token matches the potential bot
+		token, err = client.GetToken(ctx, c.tokenID)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return trace.NotFound("token with name %q not found, create the token or do not set TokenName: %v",
+					c.tokenID, err)
+			}
+			return trace.Wrap(err)
+		}
+		if !token.GetRoles().Include(types.RoleBot) {
+			return trace.BadParameter("token %q is not valid for role %q",
+				c.tokenID, types.RoleBot)
+		}
+		if token.GetBotName() != c.botName {
+			return trace.BadParameter("token %q is valid for bot with name %q, not %q",
+				c.tokenID, token.GetBotName(), c.botName)
+		}
+	}
+
+	bot := &machineidv1pb.Bot{
+		Metadata: &headerv1.Metadata{
+			Name: c.botName,
+		},
+		Spec: &machineidv1pb.BotSpec{
+			Roles: roles,
+			Traits: []*machineidv1pb.Trait{
+				{
+					Name:   constants.TraitLogins,
+					Values: flattenSlice(c.allowedLogins),
+				},
+			},
+		},
+	}
+
+	bot, err = client.BotServiceClient().CreateBot(ctx, &machineidv1pb.CreateBotRequest{
+		Bot: bot,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if c.format == teleport.JSON {
+		out, err := json.MarshalIndent(bot, "", "  ")
+		if err != nil {
+			return trace.Wrap(err, "failed to marshal CreateBot response")
+		}
+
+		fmt.Println(string(out))
+		return nil
+	}
```

Comment:
One breaking change here, previously we would output the token and join method in a machine readable way which was useful for creating bots via script:
```
$ tctl bots add test --roles=access --format=json
{
  "user_name": "bot-test",
  "role_name": "bot-test",
  "token_id": "foo",
  "ttl": 3599994659626,
  "join_method": "token"
}
```

Now we just dump the bot object, which does not have any info about the join token or method:
```
$ tctl bots add test --roles=access --format=json                                                                                                                                                                                               
{
  "kind": "bot",
  "version": "v1",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "traits": [
      {
        "name": "logins"
      }
    ]
  },
  "status": {
    "user_name": "bot-test",
    "role_name": "bot-test"
  }
}
```

I don't know whether or not anyone used this - and it mostly only applies to static tokens which are a pain - but there's not a great substitute for creating bots via script now since any auto generated tokens get eaten.

(I'm not super familiar with the operator's embedded tbot but it might be worth making sure it doesn't depend on tctl's output, since it uses token based joining)

Reply:
Good catch, I'll modify this to preserve the old JSON

Reply:
[d552a14](https://github.com/gravitational/teleport/pull/35441/commits/d552a14533cd3e5311bba8df6ce72fb7adc11553)